srml/contract/src/lib.rs

```diff
@@ -14,41 +14,68 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate. If not, see <http://www.gnu.org/licenses/>.
 
-//! Smart-contract module for runtime; Allows deployment and execution of smart-contracts
-//! expressed in WebAssembly.
+//! # Contract Module
+//! 
+//! The contract module provides functionality for the runtime to deploy and execute smart-contracts expressed in WebAssembly.
+//! To add this functionality, you need to add the contract module to `construct_runtime!` in the runtime module.
+//! The supported dispatchable functions are documented as part of the [`Call`] enum.
+//! 
+//! ## Overview
+//! 
+//! This module extends accounts with the ability to create smart-contracts and make calls to the contracts.
+//! Each smart-contract is associated with an account, which handles the code and storage. When a smart-contract is called, its associated code gets executed.
```

Comment:
good pt, we should point out the optimization here. pls correct me if im wrong @pepyakin : 

The optimization comes from not having to store the contract *creation* code. But, each new smart contract instance still has to store the logic part of the code, within its own context. i..e every smart-contract type of account has to have associated code.

Open to suggestions on how to rephrase this!

Reply:
> I feel that this line does not correctly illustrate the separation between where the code live, and where a code is "initialized". For example, if you instantiate hundreds of the same contract from the same codehash, it should not take up 100x of the contract space, because each account does not get a copy of the code...

You got it right, @shawntabrizi.

> The optimization ... 

I wouldn't call it an optimization since it is, well, how the things work here. Even if it is an optimization compared to an existing environments such as Ethereum, but not sure if we can or want to assume that the reader is familiar with Ethereum.

> ... comes from not having to store the contract creation code. But, each new smart contract instance still has to store the logic part of the code, within its own context.

I am not sure if we are on the same page here, tbh. Let's recap

- `put_code(code: &[u8])` takes code, validates it and saves it in the `code_cache`.
- Each contract have an attribute called `codehash`. This is the means how code is associated with a contract.
- To execute a contract, it's code is loaded from `code_cache` by `code_hash` and then this code is executed.